Allow file or URI arguments to metaschema-cli

aj-stein-nist opened the pull request:
# Committer Notes

Closes #297.

### All Submissions:

- [x] Have you followed the guidelines in our [Contributing](https://github.com/usnistgov/metaschema/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/usnistgov/metaschema/pulls) for the same update/change?
- [x] Do all automated CI/CD checks pass?
- ~Have you squashed any non-relevant commits and commit messages? \[[instructions](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History)\]~ Leaving for final review, to be squashed or rebased on merge.

### Changes to Core Features:

- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your core changes, as applicable?
- [x] Have you included examples of how to use your new feature(s)? Tests sufficiently demonstrate the updated usage of the tool with relevant CLI arguments and parameters.
- ~Have you updated all website](https://pages.nist.gov/metaschema) and readme documentation affected by the changes you made? Changes to the website can be made in the website/content directory of your branch.~

aj-stein-nist committed:
Add UriUtils class for URI/Path translation

Currently, we have the metaschema-cli only support local file paths for
arguments for content inputs and outputs. This loader will support the
use of MetaschemaLoader.load() with URIs and unify local file path and
URL access (primarily https://, http://) with the same command line
tooling for similar remote and local access.

Co-Authored-By: David Waltermire <david.waltermire@gsa.gov>

aj-stein-nist committed:
CLI commands allow files or URIs for #297

This PR adjusts the commands and supporting databind classes and ports
them to use URIs. The commands will use the new UriUtils class and the
toUri() function to use a remote URI supported as a URL and load the
model, schema, or target file. If a local file, it will convert the file
path to valid URI.

Integration tests have been added to validate this functionality works
with the update CLI commands.

aj-stein-nist commented on core/src/main/java/gov/nist/secauto/metaschema/core/util/UriUtils.java:
```diff
@@ -0,0 +1,60 @@
+    // disable construction
+  }
+
+  public static URI toUri(@NonNull String location, @NonNull URI baseUri) throws URISyntaxException {
```
Is this a simple enough util class we do not need comments? If so, let me know in review and I can add.

david-waltermire replied:
We should always have a Javadoc.

aj-stein-nist replied:
On it.

aj-stein-nist replied:
Javadoc added in https://github.com/usnistgov/metaschema-java/pull/298/commits/6c8f8bfc5acecbcd1cd0d02a4c4e9fa5f03450d8.

aj-stein-nist commented on core/src/test/java/gov/nist/secauto/metaschema/core/util/UriUtilsTest.java:
```diff
@@ -0,0 +1,90 @@
+        add(Arguments.of("local/relative/path/is/invalid.txt", VALID));
+        add(Arguments.of("/absolute/local/path/is/invalid.txt", VALID));
+        add(Arguments.of("1;", VALID));
+        // These tests fail, they appear to be valid in Linux on GHA
+        // https://github.com/usnistgov/metaschema-java/actions/runs/7663271718/job/20885821561
+        // add(Arguments.of(":", INVALID));
+        // add(Arguments.of(">", INVALID));
+        // add(Arguments.of("<", INVALID));
+        // https://github.com/usnistgov/metaschema-java/actions/runs/7663616200/job/20886692509?pr=298
+        // add(Arguments.of(" ", INVALID));
+        // add(Arguments.of(" ?", INVALID));
```
Now that I seem to know these negative tests won't work, I can remove them. I still cannot find a negative test with a string value unless I do some odd Unicode characters if you'd like.

```suggestion
```

aj-stein-nist replied:
Removed in https://github.com/usnistgov/metaschema-java/pull/298/commits/7957c6dd0c7710ed95672369e18974847d086f3b.

aj-stein-nist commented on metaschema-cli/src/main/java/gov/nist/secauto/metaschema/cli/commands/AbstractValidateContentCommand.java:
```diff
@@ -116,38 +117,6 @@ public List<ExtraArgument> getExtraArguments() {
-    }
-
-    List<String> extraArgs = cmdLine.getArgList();
-    if (extraArgs.size() != 1) {
-      throw new InvalidArgumentException("The source to validate must be provided.");
-    }
```
Since this isn't a file check, should I restore this? That or this whole overridden function can be potentially deleted now.

david-waltermire replied:
Yes. Please put this back.

aj-stein-nist replied:
Fixed in https://github.com/usnistgov/metaschema-java/pull/298/commits/e51fc11ed999007ea6b42683398b62fbe287d031.

aj-stein-nist commented on databind/src/main/java/gov/nist/secauto/metaschema/databind/io/yaml/YamlOperations.java:
```diff
@@ -84,9 +84,9 @@ private YamlOperations() {
-    try (BufferedReader reader = Files.newBufferedReader(target.toAbsolutePath(), StandardCharsets.UTF_8)) {
-      return (Map<String, Object>) YAML_PARSER.load(reader);
+  public static Map<String, Object> parseYaml(URI target) throws IOException {
+    try (InputStream is = ObjectUtils.notNull(target.toURL().openStream())) {
```
Per our conversation on the call, you would prefer this @david-waltermire?

```suggestion
try (InputStream is = new BufferedInputStream(ObjectUtils.notNull(target.toURL().openStream()))) {
```

aj-stein-nist replied:
Changed locally and pushing up with pairing changes momentarily.

aj-stein-nist replied:
Adjusted in https://github.com/usnistgov/metaschema-java/pull/298/commits/9f5fff5b230f5a08284469acd277fb4ff1b26a9d.

aj-stein-nist committed:
Specific error handling for different URI failures

Per discussion with @david-waltermire, update the exception handling and
associated error messages to identify if a bad domain for a remote URI
is used or a valid domain is used but a bad URL.

Co-authored-by: David Waltermire <david.waltermire@gsa.gov>

aj-stein-nist committed:
Feedback: use buffered input stream for YAML URIs

aj-stein-nist committed:
Feedback: remove infeasible UriUtilsTest cases

aj-stein-nist committed:
Feedback: arg count check for AbstractValidateContentCommand

aj-stein-nist committed:
Feedback: docs for UriUtils.toUri()

aj-stein-nist commented:
@david-waltermire integrated all changes from ad-hoc pairing and review. This is ready for final review and possible merge if you are happy with it.

david-waltermire reviewed (approved):
This is awesome. Thanks so much!
